### ISSUE TITLE
vagrant adb box should have the latest docker engine

### ISSUE THREAD

**Issue by navidshaikh** [projectatomic/vagrant-service-manager#3]
_From @bamachrn on February 2, 2016 6:59_

While using vagrant-adbinfo plugin with docker client of local machine we get

$ docker ps
Error response from daemon: client is newer than server (client API version: 1.21, server API version: 1.20)

My local machine is having latest version of docker. But the engine in adb box is old one.
We should make changes in the system, so that when one installs vagrant-adbinfo plugin it should automatically update the docker engine of the vagrant box.

_Copied from original issue: projectatomic/vagrant-adbinfo#85_

**Comment by navidshaikh** [comment]
_From @mohammedzee1000 on February 2, 2016 7:8_

If stability is an issue (as we do stay behind a couple of versions behind upstream for the same reason), we could have user being recommended for downgrade or the same happening automatically on the guest, though i do prefer if the box version gets updated.

**Comment by navidshaikh** [comment]
_From @LalatenduMohanty on February 2, 2016 7:10_

@bamachrn The ADB Vagrant box is based on CentOS core i.e. it can only have RPMs which is available in CentOS. So we wont be able to go to a docker version which is not present in CentOS.  

I think we should think about how we can ship the docker client through the ADB box and then use the correct/compatible docker client version from the host to communicate

**Comment by navidshaikh** [comment]
_From @bamachrn on February 2, 2016 7:15_

@LalatenduMohanty This is good. But we should have some way to at least make the client and server versions in the same native api version. As it is not backward compatible, and user will face trouble.

We should get a way so that the user need not to think about the version management.

**Comment by navidshaikh** [comment]
_From @mohammedzee1000 on February 2, 2016 7:19_

So basically we need to figure out how to run 2 versions of docker client seamlessly on a system, or we could inform the user and with his permission, downgrade the client automatically.

**Comment by navidshaikh** [comment]
_From @LalatenduMohanty on February 2, 2016 7:27_

@bamachrn Agree with you. User should not have to think about it.

@mohammedzee1000 these things can be handled automatically we we create single click installers for ADB.  But yes, we need to find a way to do that, then automate + integrate it

**Comment by navidshaikh** [comment]
_From @mohammedzee1000 on February 2, 2016 10:57_

Few things we could try (this problem will arise especially in the dev environment than the ops one)
- If there is a version mismatch, create another vagrant vm with docker client of appropriate version installed and route docker traffic from there.(suggestion from shubam).
- Package the docker client of appropriate version with a slightly modified name 'adb-docker' (this coud be done through links as well. ie newname link pointing to packaged docker client in respective users bin directory) for example.
- Check for version mismatch, and with users permission, upgrade or downgrade accordingly.

**Comment by navidshaikh** [comment]
_From @containscafeine on February 3, 2016 9:45_

In addition to what @mohammedzee1000 has suggested,

Instead of using Docker CLI client on the base machine, can [Remote API](https://docs.docker.com/engine/reference/api/docker_remote_api/) or one of these [Remote API clients](https://docs.docker.com/engine/reference/api/remote_api_client_libraries/) be used to access Docker in the Vagrant VM? This way upgrading or downgrading Docker on the base machine does not have to be worried about, and this will also work across platforms.

However, in the case of using:

[Remote API](https://docs.docker.com/engine/reference/api/docker_remote_api/):
- Multiple API versions have to be supported.
- Effectively, this is like maintaining a new Docker Remote API client.

[Remote API client](https://docs.docker.com/engine/reference/api/remote_api_client_libraries/):
- If these are well maintained (Docker does not maintain these, except [docker-py](https://github.com/docker/docker-py)), then each API version does not have to be supported and code remains the same.
- These might not have support for all the Docker CLI client features (even [docker-py](https://github.com/docker/docker-py) does not support 'everything' which the CLI client does).

On another note, how about a hybrid approach? When the Remote API client does not work (feature not supported or version incompatibility), the Remote API can be used to do that thing.

Please correct me if I'm wrong or talking off-track!

**Comment by navidshaikh** [comment]
_From @tuxdna on February 3, 2016 10:16_

As mentioned by @containscafeine it makes sense to use Remote API.

Since we are not going just upgrade to docker version 1.9 very soon ( because it has technical debt ), we may just as well use REST API. The reason is simple, the underlying VM is really just CentOS, which has docker version 1.2.

We can build simple tools to address the use-cases ( like `docker ps` in this issue ), and provide a seamless integration with the REST API endpoints rather than depending on docker server/client version compatibility.

**Comment by navidshaikh** [comment]
_From @bamachrn on February 3, 2016 11:20_

I would like to add few things here.
1. External docker client is only required when we are exporting environment variables to the local machine.
2. In other scenarios we do not need external docker client.

So what I was having in mind is why don't we just add one more environment variable export to the existing set. This will redirect all the commands from the local machine, which in turn will use the docker client inside the adb vagrant box.

we just need a way to redirect the commands to the internal docker client this will help in all the cases including windows, mac.

I believe we should not do any changes to the users local machine.

I am working for the same to get a way to use the internal docker client. Please suggest if something else is in mind.

**Comment by navidshaikh** [comment]
_From @tuxdna on February 3, 2016 11:32_

@bamachrn The question then is "who will use that extra environment variable"? Will this be 
- the docker client OR 
- a tool that can understand this variable and forward the commands to the docker in VM

We would still need some running code ( or a tool ) that understands this extra env variable.

**Comment by navidshaikh** [comment]
_From @bamachrn on February 3, 2016 11:45_

@tuxdna extra environment variable is for the local users machine.  When user exports the docker host variable using `eval "$(vagrant adfinfo)"`, this one will also get added. So User will not see any changes in behavior. And this will work seamlessly. 
If the user wants to use vagrant ssh or eclipse, this env variable will not be exported hence it will not have any effect to users environment.

**Comment by navidshaikh** [comment]
_From @tuxdna on February 3, 2016 11:53_

@bamachrn that doesn't address the question that which running code ( or tool ) will actually understand this extra environment variable.

Perhaps you are saying that this environment variable might indicate to the tool to forward requests over SSH. That will work, but the question is what code is going to do that? Obviously docker version 1.9 doesn't do it at the moment.

**Comment by navidshaikh** [comment]
_From @bamachrn on February 3, 2016 12:4_

@tuxdna yes I am working for the same to find out a way to do that.

**Comment by navidshaikh** [comment]
_From @tuxdna on February 3, 2016 12:22_

@bamachrn great! You can put your ideas here, or let me know we can collaborate on it.

**Comment by navidshaikh** [comment]
_From @hferentschik on February 5, 2016 9:3_

Is this not a similar discussion to the oc client? I think the optimal solution would be, if the VM could contain the matching Docker and OpenShift CLI (for all OS types) and adbinfo (resp. its successor service-manager) allows to copy them out to the host. The plugin cloud in this case also detect the OS type automatically. For example:

```
$ vagrant service-manager docker copy-client
```

and

```
$ vagrant service-manager openshift copy-client
```

The user can then decide whether he wants to execute the CLI locally (from the same directory) like `./docker` or whether he wants to update its `PATH` temporarily/permanently.

**Comment by navidshaikh** [comment]
_From @praveenkumar on February 8, 2016 4:44_

@hferentschik Yes we can cache oc and docker binary to ADB/CDK box but it might be using a rpm because build system don't have external network connectivity. I will discuss with @LalatenduMohanty  and find out what would be a good solution for cache those binaries.

**Comment by navidshaikh** [comment]
> find out what would be a good solution for cache those binaries.

How about including the binaries in adb-utils?

**Comment by navidshaikh** [comment]
_From @praveenkumar on February 8, 2016 10:54_

> How about including the binaries in adb-utils?

Adding binary into RPM package which something I avoid because it violate RPM packaging Guideline.

https://fedoraproject.org/wiki/Packaging:Guidelines#No_inclusion_of_pre-built_binaries_or_libraries

**Comment by navidshaikh** [comment]
_From @hferentschik on February 8, 2016 14:16_

where is rpm involved here?

**Comment by navidshaikh** [comment]
_From @praveenkumar on February 8, 2016 14:50_

> where is rpm involved here?

How do you suggest to include those binaries then?

**Comment by tuxdna** [comment]
Do we really need to add binaries at all? How about we expose a bash function in current shell something like this:

```
$ function docker() { vagrant ssh -c "docker $@"; }
```

And then use the function instead of the `docker` binary:

```
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
Connection to 192.168.121.243 closed.

```

We can do this conditionally, i.e. only if the docker versions between host and guest mismatch.

**Comment by hferentschik** [comment]
Bash function is great, but it won't work for the majority of our target audience which is Windows. Also I am not sure how well this works when the docker commands get more complicated and in particular contain volume mounts. You would specify directories on your host, but the command gets executed in the guest where the directory won't exists (at least not under the path you specified).

**Comment by hferentschik** [comment]
> How do you suggest to include those binaries then? 

Well, I would have them placed in specific locations during the build of the ADB/CDK. The service-manager would then just copy from there. Not sure how this would be possible with the current way the VM is put together and whether this violates other build requirements. 

If we cannot add the binaries to the box, we maybe could write a script which retrieves the binaries on first request and then caches them. Obviously this means that one needs a internet connection on the first call, but the script could cater for that. Maybe this would help to get around constraints related on what one can and cannot do when building the VM.

**Comment by tuxdna** [comment]
@hferentschik I agree with both of your points.

Bash functions would be good for only inspecting the state of containers from host to guest.  As you described, when make some changes from host to gets,  it will cause confusion to the end-user. However, I think this issue reported by @bamachrn is specific to Linux  ( hosts ).

For compatible Docker binary, lets use a script to download appropriate version [1]. Now this will work well for Linux or Mac, but for Windows it could be more involved ( different shell, directory structure, installation mechanism etc. )

[1] https://docs.docker.com/engine/installation/binaries/

**Comment by bexelbie** [comment]
We are limited here by what is shipping in a particular distros build of the ADB.  I believe we should close this issue.  We are considering #90 which may fix this.  But in the end, this can be solved by a documentation patch #136

**Comment by navidshaikh** [comment]
> We are limited here by what is shipping in a particular distros build of the ADB.

I agree.
